cmd/hub/mcscontroller/controller.go

```diff
@@ -118,15 +121,205 @@ func (r *ServiceExportReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 //	 automatically deleted by k8s when their ServiceExport is deleted.
 //
 func (r *ServiceExportReconciler) reconcile(ctx context.Context, name types.NamespacedName) (ctrl.Result, error) {
-	_ = r.Log.WithValues("service", name)
+	log := r.Log.WithValues("service", name)
+	//1:
+	var exportList mcsv1a1.ServiceExportList
+	if err := r.Client.List(ctx, &exportList, client.MatchingLabels{"name": name.Name, "namespace": name.Namespace}); err != nil {
+		log.Error(err, "unable to list service's service export")
+		return ctrl.Result{}, err
+	}
+	//2:
+	primaryEs, err := getPrimaryExportObject(exportList)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	//3:
+	for _, se := range exportList.Items {
+		if se.DeletionTimestamp != nil { //3a:
+			continue
+		}
 
-	//------------------------------------------
-	// -- @TODO CODE ABOVE SPECIFIED BEHAVIOR --
-	//------------------------------------------
+		exportSpec := &mcs.ExportSpec{}
+		err := exportSpec.UnmarshalObjectMeta(&se.ObjectMeta)
+		if err != nil {
+			log.Info("Failed to unmarshal exportSpec from serviceExport")
+			return ctrl.Result{}, err
+		}
 
+		err = exportSpec.IsCompatibleWith(primaryEs)
+		//3b:
+		if err != nil {
```

Comment:
On success, err will be `(*mcs.CompatibilityError)(nil)` so the wrong path will be taken. You either need to test for `(*mcs.CompatibilityError)(nil)` or change IsCompatibleWith to return bare error instead of `*CompatibilityError`(I think)

Reply:
https://go.dev/play/p/iACOCIEeVgO ?

Reply:
```
      runtime.sigpanic (signal_unix.go:734)
      github.com/submariner-io/lighthouse/pkg/mcs.(*CompatibilityError).Error (exportspec.go:78)
      main.(*ServiceExportReconciler).reconcile (controller.go:154)
      main.(*ServiceExportReconciler).Reconcile (controller.go:90)
```
In my code that is the last line here:
```
		err = exportSpec.IsCompatibleWith(primaryEs)
		//3b:
		if err != nil {
			r.updateServiceExportConditions(ctx, &se, mcsv1a1.ServiceExportConflict, corev1.ConditionTrue, "", err.Error())
```
Fixed by:
```
		if err != (*mcs.CompatibilityError)(nil) {
```

Reply:
The issue seems to be caused by: `err` is initially assigned a value from a function that returns `error` then it gets reassigned a value from a function that returns `*CompatibilityError`.

https://go.dev/play/p/wl8tXeK1xf7

Reply:
I see.
So `IsCompatibleWith(...) error` instead of `IsCompatibleWith(...) *CompatibilityError`?

Reply:
It seems like  'err != (*mcs.CompatibilityError)(nil)' is working without giving up on the custom err struct '*CompatibilityError' .

Reply:
you can always "type assert" the returned error interface (e.g., if you needed access to some specific methods, but it seems that `Error()` returns the needed information so unclear if we need `Cause()` at all...). I believe returning `error` would be more idiomatic Go. 
@donaldh wdyt?

Reply:
Yes, it’s more idiomatic to return error. The issue is covered here:

https://go.dev/doc/faq#nil_error

Reply:
@davidohana